platform/default/src/mbgl/gl/headless_backend.cpp

```diff
@@ -43,10 +43,19 @@ HeadlessBackend::HeadlessBackend(const Size size_,
 
 HeadlessBackend::~HeadlessBackend() {
     gfx::BackendScope guard{ *this, gfx::BackendScope::ScopeType::Implicit };
+    // Some implementations require active context for GL functions to work
+    if (impl->glNeedsActiveContext()) {
+        impl->activateContext();
+    }
```

Comment:
We should check if `impl` is valid.
The name `glNeedsActiveContext` is too generic, maybe we should be more specific in naming this.
But it will be better if possible to fix it outside of this class. What is the callstack, from where the destructor is called?

Reply:
Will add the check, nice spot!

In the MBGL context below some GL functions are called on destruction (when `context.reset()` is called). The problem is that Qt GL context is not valid when deactivated so it segfaults.

One option was to add if clauses to every GL call in the wrapper, but maybe that's an overkill (and also potential slowdown), so by advice of Bart I found an easier workaround.

Regarding the naming I think the name fits quite well, maybe `glCallsNeedActiveContext`?

Reply:
Regarding the naming, I mean `glNeedsActiveContextOnDestruction`.

This workaround adds more code to core which is needed on single platform, so makes the code harder to read. It assumes also the `context.reset` calls some gl functions, which can change in the future and this check will remain here.

I was thinking more from where the destructor is called, maybe there is the issue, why the destructor is called and context is no more active?

Reply:
I see your concerns. Maybe we can open a ticket to try to improve it in the future.

It's called in the HeadlessBackend destructor.

My bigger question is why platform contexts are not inside the core MBGL context but somehow separate. This does not make much sense to me, but it would be a larger development to improve (and probably something to consider with your modularisation work).

Also I'm not sure what happens in other platforms. Do GL calls just take some default context then?

Reply:
I mean from where the `HeadlessBackend destructor` is called.

Reply:
Each unit test that renders basically.

Reply:
So maybe there the fix should happen. Maybe there is a good place to make sure the destructor is called with an active context, or a default one. When I will have time I will check on iOS how the context is handled in this scenario.

Reply:
Yes, I think the whole headless rendering needs a bit of rework. Let me open a ticket and we merge it as is (will rename the property and add impl check first).